### ISSUE TITLE
Primitive iterator looses sync after too many elements are removed

### ISSUE THREAD

**Issue by alexo134** [eclipse/eclipse-collections#481]
Consider this code:

```.java
MutableIntSet s = IntSets.mutable.of();

int max = 100000;

for(int i = 0; i < max; i++) {
    s.add(i);
}

MutableIntIterator iterator = s.intIterator();

while (iterator.hasNext()) {
    // remove 2/3 of elements
    if (iterator.next() < ( max * 2 / 3)) {
        iterator.remove();
    }
}
```
this will cause this exception:
```
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 262144
	at org.eclipse.collections.impl.set.mutable.primitive.IntHashSet$InternalIntIterator.next(IntHashSet.java:1624)
```
It works for smaller numbers of elements like max = 1k or 10k but starts to fail for 100k and more.

And I found the cause: 
```InternalIntIterator.remove()``` calls the standard methods ```IntHashSet.remove();``` which does rehashing when there is too many entries occupied with sentinels.

```.java
...
if (this.occupiedWithSentinels > this.maxOccupiedWithSentinels()) {
    this.rehash();
}
...
```
and this makes iterator invalid.

**Comment by nikhilnanivadekar** [comment]
@alexo134 thanks for reporting this! 
We will take a look and come back on this issue.

**Comment by nikhilnanivadekar** [comment]
@alexo134 raised a #484 to fix it. Once merged it will be available with 9.2.0 release which is slated for 04/18: https://projects.eclipse.org/projects/technology.collections/releases/9.2.0
